cwltool/workflow.py

```diff
@@ -459,6 +511,84 @@ def visit(self, op):
             s.visit(op)
 
 
+def static_checker(workflow_inputs, workflow_outputs, step_inputs, step_outputs):
+    # type: (List[Dict[Text, Any]], List[Dict[Text, Any]], List[Dict[Text, Any]], List[Dict[Text, Any]]) -> None
+    """Check if all source and sink types of a workflow are compatible before run time.
+    """
+
+    # source parameters: workflow_inputs and step_outputs
+    # sink parameters: step_inputs and workflow_outputs
+
+    # make a dictionary of source parameters, indexed by the "id" field
+    src_parms = workflow_inputs + step_outputs
+    src_dict = {}
+    for parm in src_parms:
+        src_dict[parm["id"]] = parm
+
+    step_inputs_val = check_all_types(src_dict, step_inputs, "source")
+    workflow_outputs_val = check_all_types(src_dict, workflow_outputs, "outputSource")
+
+    warnings = step_inputs_val["warning"] + workflow_outputs_val["warning"]
+    exceptions = step_inputs_val["exception"] + workflow_outputs_val["exception"]
+
+    warning_msgs = []; exception_msgs = []
+    for warning in warnings:
+        src = warning.src; sink = warning.sink; linkMerge = warning.linkMerge
+        msg = ("Warning: potential type mismatch between source '%s' (%s) and "
```

Comment:
How do we include `SourceLine(sink).makeError()` in it as well? Maybe `SourceLine(src).makeError(SourceLine(sink).makeError(...))`?